File: 2. Add Two Numbers/addTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode root = new ListNode(0); // fist position of the linked list and set as the root
        ListNode newList = root;
        int extraNum = 0; // to deal with the condition that sum of two number is bigger than 9
        while(l1 != null || l2 != null || extraNum != 0) {
            // take example 3 as a special one
            //  l1 [9,9,9,9,9,9,9]
            //  l2 [9,9,9,9]
            // sol [8,9,9,9,0,0,0,1]
            int l1Val = l1 != null ? l1.val : 0;
            int l2Val = l2 != null ? l2.val : 0;
            int sumNum = l1Val + l2Val + extraNum;
            extraNum = sumNum / 10;
            ListNode sumNode = new ListNode(sumNum % 10);
            newList.next = sumNode;
            newList = sumNode;

            if(l1 != null)
                l1 = l1.next;
            if(l2 != null)
                l2 = l2.next;
        }
        return root.next;
    }
}

File: 2. Add Two Numbers/addTwoNumbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummyNode = ListNode(0)
        temp = dummyNode
        outRange = 0
        while l1 or l2:
            l1Val = 0 if l1 is None else l1.val
            l2Val = 0 if l2 is None else l2.val

            sumTemp = l1Val+l2Val+outRange
            store = sumTemp % 10
            outRange = sumTemp // 10
            
            temp.next = ListNode(store)
            temp = temp.next

            if l1: l1 = l1.next
            if l2: l2 = l2.next
        
        if outRange != 0: temp.next = ListNode(outRange)

        return dummyNode.next
